### ISSUE TITLE
ReplicationController is too slow to support 2000-node clusters

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#25563]
After changing kubemar-scale to run 2000-node kubemarks (with protobufs), it is failing with errors like these:

14:38:23   [Feature:Performance] should be able to handle 30 pods per node [It]
14:38:23   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/load.go:174
14:38:23 
14:38:23   scaling rc load-small-rc-3813 for the first time
14:38:23   Expected error:
14:38:23       <*errors.errorString | 0xc8216acd70>: {
14:38:23           s: "error while scaling RC load-small-rc-3813 to 2 replicas: timed out waiting for the condition",
14:38:23       }
14:38:23       error while scaling RC load-small-rc-3813 to 2 replicas: timed out waiting for the condition
14:38:23   not to have occurred

@davidopp @gmarek @fgrzadkowski @kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
I did some more debugging around that and I think this is more a test problem than system problem.

Basically, when looking into logs we see that it takes even minutes for a test to observe some pod creation.
Whereas in the system this pod was already created, scheduled and started in the meantime.

I have a feeling that the test itself is cpu-starved on Jenkins. (though I don' understand why 1000-node kubemark was passing, because the load per second is supposed to be roughly the same).

**Comment by wojtek-t** [comment]
I will try to run this test from my desktop to see if I see the same.

**Comment by wojtek-t** [comment]
Actually - I was completely wrong.

I did some deeper debugging, and it seems it's not test, but ReplicationController that is slow.

As an example in this run:
https://storage.googleapis.com/kubernetes-jenkins/logs/kubernetes-kubemark-gce-scale/1148/build-log.txt

see ReplicationController: "load-small-rc-3813"

- It took 3+ minutes to create those pods in the system
- Once they were already created, they were all running within <10s

So in general throughput of apiserver & scheduler are good enough, but
ReplicactionController is too slow.

**Comment by davidopp** [comment]
Do you think it is rate limits again?

cc/ @bprashanth who is always interested in all things ReplicationController-related :)

**Comment by wojtek-t** [comment]
I think it's something different this time. I have a hypothesis that this might be related to shared pod informer, but this is just my guess now. I will try to debug this today.

**Comment by mqliang** [comment]
@wojtek-t 

I think the reason may be: `getPodJob` and `getPodPetSet`, `getPodReplicaSet` are still slow. I have send several PRs to add lookup cache for RC/RS/DS, but the lookup cache for Job/PetSet has not been merged yet. The PR trying to add namespace index for RC/Pod  has been merged, but I am too busy to send PRs for RS/Job/DS/Deployment.

And since we use shared pod informer, pod local cache will becomes ready only after ALL Job/DS/RC/RS/PetSet/Deployment has been put into their corresponding workqueue, replication_controller will NOT begin manage replicas before pod local cache becomes ready, see [podStoreSynced] (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L537-L543).

**Comment by wojtek-t** [comment]
I don't think it's about namespace, because we were always using a single namespace in a test (we want to fix it at some point, but we are not there yet).

However, when I'm looking into logs, the problem is that we are not aware of some pods, or sth like that. The problem is that we even start syncing a given RC really late. 

The podStoreSynced may be a problem if the controller is restarted - but once it is initialed, this will always be true (and this is the case in our tests). So it's definitely not about initialization (because we start processing the first RC immediately), but the issue is that we are lagging more and more over time.

The the problem may potentially be lock contention on podStore.

**Comment by wojtek-t** [comment]
Also note, that we don't have any Jobs,PetSets, DS, etc. in our tests (just RCs). SO no matter what the implementation is, all those get* methods should be very fast (in this particular test).

**Comment by mqliang** [comment]
@wojtek-t Ok. 

> but the issue is that we are lagging more and more over time. The the problem may potentially be lock contention on podStore.

Another possible reason is: periodic relist/resync saturate work queue(if you run the test for a long time or several times, and it becomes worse and worse). Just my guess thought, hoping it could help you debug.

**Comment by mqliang** [comment]
BTW, as for lock contention on podStore. The [Big lock](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/thread_safe_store.go#L53) for threadSafeMap make me feel nervous, we could use the locking mechanism which is used in concurrentHashMap in java: breaks the contents up into Segments to avoid locking the whole map. 

Some high-quality implementation of concurrentHashMap for Go may available on github.

**Comment by wojtek-t** [comment]
@mqliang - yes, that's exactly what I meant by lock contention on podStore.

**Comment by wojtek-t** [comment]
Since it is actually a matter of system throughput in general, for 1.3 we will probably change our tests to use more than a single namespace (which should singnificantly improve things). Note that if someone is starting 1000s of RCs and 60000 pods in a cluster, we can probably assume he/she won't be doing it in a single namespace.

However, we should also solve the underlying problem, but it seems that it's not P0 for 1.3. And controllerRef should help a lot with that, and it should be ready for 1.4

**Comment by wojtek-t** [comment]
I added some traces and run the test locally, and lock contention on podStore doesn't seem to be a problem. The biggest problem is that we are resyncing each RC periodically every 30s. And if we have say 3000RC in a queue for periodic resync and we are creating a new RC, it just takes time to get to processing the new one.
I think I already have some ideas how to fix it and will look into it deeper next week.

**Comment by mqliang** [comment]
> The biggest problem is that we are resyncing each RC periodically every 30s. And if we have say 3000RC in a queue for periodic resync and we are creating a new RC, it just takes time to get to processing the new one.

In other words, it's because periodic relist/resync saturating the work queue. Then https://github.com/kubernetes/kubernetes/issues/24375 is related.

**Comment by wojtek-t** [comment]
@mqliang agree.
However, we should try to avoid frequent periodic lists, instead of adding complicated mechanisms of priorities. I need to verify if what I'm thinking works as I want and discuss it, but I think we should be able to solve it in an easier way than adding priorities.

**Comment by mqliang** [comment]
@wojtek-t priority mechanism may not be so complicated as you may think, we could decorate the existing code to implement one. 

One possible implementation sketch could be https://github.com/mqliang/kubernetes/commit/d0aea9db2da4285049b05cc0256e44fbab83cc74 , where:

* add SyncAwareInformer which could distinguish Sync events from Update events, leaving the previous Informer as it is, since it has been widely used by many clients outside the core kube repo
* decorate the existing work queue to implement a priority-aware work queue

Of course, if anybody has a better idea to solve the issue, I am fine with not adding priority.

**Comment by mqliang** [comment]
And some other possible solution I can think of:

1) Do not use priority, use ***two*** work queue instead, one for Add/Update/Delete events and another for Sync events.

2) As for the informer, we can also leave the Informer as it is, but do the reflect.DeepEqual at the [UpdateFunc](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/replication_controller.go#L147). If RC unchanged, it's a Sync event. Then put the RC into work queue with a relatively lower priority(or into the workqueue dedicated for Sync events)

**Comment by wojtek-t** [comment]
@mqliang - I did some experiments and had some internal conversation and I'm going to do short term (Mon) is:
- explicitly requeue controller for resync in case of errors (creating, deleting pods, updating RC status, etc.)
- decrease the frequency of full resync - it should basically be not needed then

Those changes are really easy, and should help significantly with overall system throughput.

Since we will start code freeze in week from now, there is no time to implement priority mechanism. We can get back to it in 1.4 timeframe.

**Comment by wojtek-t** [comment]
OK - those fixes seem to be enough for 1.3. So I'm going to have a more proper fix outside 1.3

**Comment by mqliang** [comment]
After more thinking, the priority mechanism have starvation problem: a event with poor priority may will be never processed.

relist/resync shouldn't be needed at all in a healthy system. But I think we may need it since it is a protection against disappearing objects and watch notification(may should not happen at all)

If we finally decide compeletely get rid of relist/resync, then the problem doesn't exist any more. If we want it as a protection mechanism, I think we may could use multi workqueue, instead of a single one: use a dedicate workqueue and several dedicate working routes for Sync events.

implementation sketch could be https://github.com/mqliang/kubernetes/commit/64208382ab5802a18339fa5c32505c91dabd1060

@deads2k FYI

**Comment by mqliang** [comment]
One problem of multi-workqueue is: there is contention if a rc is both in the workqueue for Add/Update/Delete events and the workqueue for Sync events.

We should be careful about the duplication removal.

**Comment by davidopp** [comment]
Moving to 1.4 milestone based on this comment
https://github.com/kubernetes/kubernetes/issues/25563#issuecomment-220365798

**Comment by hongchaodeng** [comment]
@wojtek-t 
How can I reproduce?

**Comment by mqliang** [comment]
@hongchaodeng You could reproduce it by reverting this PR: https://github.com/kubernetes/kubernetes/issues/25648. The problem here is: a large number of Sync events(produced by periodic resync) will saturate workqueue, see https://github.com/kubernetes/kubernetes/issues/24375 for more details. 

To solve the fundamental problem, I propose to use multi-work queue: one workqueue for CRUD events(usually triggered by user operations), and one workqueue for Sync events(produced by periodic resync). A initial implementation is https://github.com/kubernetes/kubernetes/pull/27165

**Comment by hongchaodeng** [comment]
@mqliang Thanks for writing out the code. I would speak highly of this behavior and code is our best diplomat. Highly recommended!

Back to the issue.

> produced by periodic resync

Why do we need resync? Why RC needs to resync?
If we can fix informer to never lose events, can we get rid of resync? If so, we probably don't need to optimize it and it will have high performance gain as well.

**Comment by mqliang** [comment]
@hongchaodeng 
> If we can fix informer to never lose events, can we get rid of resync? If so, we probably don't need to optimize it and it will have high performance gain as well.

We want resync because it's able to handle delta store comparisons against an ***external*** resource.  I.e. if you have X external resources, and N internal resources, you want to verify X=N every
few minutes.(Note: informer was widely used by some external repo). Since we can not receive events about external resource, we need periodic check. For example, when ServiceController watch a ADD event about a service, it will create a LB on cloud provider, right? But if you delete those LB on cloud provider, service will not get any events. So, ServiceController need periodic resync to ensure the LB on  cloud provider was in sync with services in k8s.

Just my personal opinion, @wojtek-t may could provide other

**Comment by hongchaodeng** [comment]
> if you delete those LB on cloud provider, ServiceController will not get any events. So, ServiceController need periodic resync to ensure the LB on cloud provider was in sync with services in k8s.

Without resync, there are two solutions for it:
- service controller can watch the LB changes
- There could be another worker thread that periodically iterate all services in local cache (from informer), and check whether corresponding LB still exists or not.

Either solution in above will help decouple the logic and help reduce complexity of each part. Informer will be simpler. Service controller doesn't need to understand what's Resync from informer and can better manage its code.

**Comment by wojtek-t** [comment]
Yes - what @mqliang wrote is basically the reason. For RC ideally we don't need resync at all, but there are controllers which need to periodically scan all objects to ensure that everything is ok. LBs are one example, PV-controller is another one.

@hongchaodeng - regarding your suggestions:
- service controller cannot watch LB changes, because cloud providers don't implement watch in general (and LB is not a k8s object in this example)
- adding another worker is much more complicated, because you need to implement it in all controller that need it;
 If we have periodic resync, controllers that need it can switch it on, and those that don't need can opt out. That means that this is only a configuration bit for a controller.
It's much better to avoid code duplication (which in fact it would be if we decide to add workers in controllers)

**Comment by hongchaodeng** [comment]
> adding another worker is much more complicated, because you need to implement it in all controller that need it; ... It's much better to avoid code duplication

It could be achieved very easily using a common routine.
From user's perspective and informer implementation perspective, we should separate such "iterating items" events from informer callbacks AddFunc, etc. Why are we coupling them inside informer? Decoupling them is good for debugging, maintaining.

Why wouldn't I write out the code and do some real comparison? I will submit some code shortly.

**Comment by timothysc** [comment]
OK, this issue is kind of conflated on a re-read-through.  

- RC's don't really need a rsync (I think we can agree on that), however we've beaten on this with the latest code and we can see that with number of RC's it's not so bad anymore with the re-list removed. 

Totally a +1 on setting resync interval to 0 there. 

@mqliang are you testing the latest tip of master?
What else seems to be the concern?

**Comment by mqliang** [comment]
@timothysc Ok, I am fine with setting resync interval as 0 for RC. But for ServiceController and PV-controller which we do need resync, the work queue saturation problem still exists.

**Comment by timothysc** [comment]
@rrati can pull the trigger on resync updates, reference this ticket.

**Comment by goltermann** [comment]
@wojtek-t @rrati @timothysc any more work needed for 1.4 on this one?

**Comment by wojtek-t** [comment]
We are good for 1.4 - moving to 1.5.

**Comment by wojtek-t** [comment]
This issue is completely obsolete.
